soc/xtensa/intel_adsp/common/ipc.c

```diff
@@ -205,12 +208,110 @@ static int dt_init(const struct device *dev)
 	return intel_adsp_ipc_init(dev);
 }
 
+#ifdef CONFIG_PM_DEVICE
+
+void intel_adsp_ipc_set_resume_handler(const struct device *dev,
+	intel_adsp_ipc_resume_handler_t fn, void *arg)
+{
+	struct ipc_control_driver_api *api =
+		(struct ipc_control_driver_api *)dev->api;
+	struct intel_adsp_ipc_data *devdata = dev->data;
+	k_spinlock_key_t key = k_spin_lock(&devdata->lock);
+
+	api->resume_fn = fn;
+	api->resume_fn_args = arg;
+
+	k_spin_unlock(&devdata->lock, key);
+}
+
+void intel_adsp_ipc_set_suspend_handler(const struct device *dev,
+	intel_adsp_ipc_suspend_handler_t fn, void *arg)
+{
+	struct ipc_control_driver_api *api =
+		(struct ipc_control_driver_api *)dev->api;
+	struct intel_adsp_ipc_data *devdata = dev->data;
+	k_spinlock_key_t key = k_spin_lock(&devdata->lock);
+
+	api->suspend_fn = fn;
+	api->suspend_fn_args = arg;
+
+	k_spin_unlock(&devdata->lock, key);
+}
+
+/**
+ * @brief Manages IPC driver power state change.
+ *
+ * @param dev IPC device.
+ * @param action Power state to be changed to.
+ * @return int Returns 0 on success or optionaly error code from the
+ * registered ipc_power_control_api callbacks.
+ */
+static int ipc_pm_action(const struct device *dev, enum pm_device_action action)
+{
+	int ret = 0;
+	const struct ipc_control_driver_api *api =
+		(const struct ipc_control_driver_api *)dev->api;
+
+	/* All device pm is handled during resume and suspend */
+	switch (action) {
+	case PM_DEVICE_ACTION_SUSPEND:
+		pm_device_state_lock(dev);
+		pm_device_busy_set(dev);
```

Comment:
Why locking the current state and set it as busy ?

Reply:
As of `pm_device_state_lock` usage, please see my https://github.com/zephyrproject-rtos/zephyr/pull/56168#discussion_r1171105609 answer.
However, good catch on the `pm_device_busy_set()`. Looks like I misused this mechanism as it has other purpose that I though. Turns out it is used to **prevent** power state changes when eg. hardware operation is performed.
From the [Device Power Manager - Busy indication](https://docs.zephyrproject.org/3.2.0/services/pm/device.html#busy-status-indication):

> (...) Suspending a device which is in the middle of a hardware transaction, such as writing to a flash memory, may lead to undefined behavior or inconsistent states. This API guards such transactions by indicating to the kernel that the device is in the middle of an operation and should not be suspended.

I'll remove this busy indicator (its purpose is different). Thanks.

Reply:
Fixed, implementation changed. Moved Device busy indicators to IPC transactions where registers are written.